data_science/archive_exploration/01 - subject coocurrence.ipynb

```diff
@@ -0,0 +1,386 @@
+    " ['Heart Diseases', 'Disease'],\n",
+    " ['Cardiology', 'Disease'],\n",
+    " ['Disease', 'Heart'],\n",
+    " ['Heart', 'Heart'],\n",
+    " ['Heart Diseases', 'Heart'],\n",
+    " ['Cardiology', 'Heart'],\n",
+    " ['Disease', 'Heart Diseases'],\n",
+    " ['Heart', 'Heart Diseases'],\n",
+    " ['Heart Diseases', 'Heart Diseases'],\n",
+    " ['Cardiology', 'Heart Diseases'],\n",
+    " ['Disease', 'Cardiology'],\n",
+    " ['Heart', 'Cardiology'],\n",
+    " ['Heart Diseases', 'Cardiology'],\n",
+    " ['Cardiology', 'Cardiology']]\n",
+    "`\n",
+    "\n",
+    "The `cartesian()` function which I've defined above will do that for us. We then find the appropriate intersection in the matrix and add another unit of 'strength' to it.  \n",
+    "We'll do this for every row of subjects in the `['Subjects']` column."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for row_of_subjects in tqdm(df['Subject'].dropna()):\n",
+    "    for subject_pair in cartesian(row_of_subjects, row_of_subjects):\n",
+    "        subject_index_1 = subject_to_index[clean(subject_pair[0])]\n",
+    "        subject_index_2 = subject_to_index[clean(subject_pair[1])]\n",
+    "\n",
+    "        adjacency[subject_index_1, subject_index_2] += 1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can do all sorts of fun stuff now - adjacency matrices are the foundation on which all of graph theory is built. However, because it's a bit more interesting, I'm going to start with some dimensionality reduction. We'll get to the graphy stuff later.  \n",
+    "Using [UMAP](https://github.com/lmcinnes/umap), we can squash the $n \\times n$ dimensional matrix down into a $n \\times m$ dimensional one, where $m$ is some arbitrary integer. Setting $m$ to 2 will then allow us to plot each subject as a point on a two dimensional plane. UMAP will try to preserve the 'distances' between subjects - in this case, that means that related or topically similar subjects will end up clustered together, and different subjects will move apart."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "embedding_2d = pd.DataFrame(UMAP(n_components=2)\n",
+    "                            .fit_transform(adjacency))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "embedding_2d.plot.scatter(x=0, y=1);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can isolate the clusters we've found above using a number of different methods - `scikit-learn` provides easy access to some very powerful algorithms. Here I'll use a technique called _agglomerative clustering_, and make a guess that 15 is an appropriate number of clusters to look for."
```

Comment:
Is 15 a good choice in general, or is that specific to this problem?

Reply:
Intuition based on the kind of dataset we're working with, past experience with that particular clustering algorithm, the way that the data is distributed under UMAP etc. In other contexts one might choose 5, 30, 200...

Reply:
This should be noted in the document.